test/session_test.dart

```diff
@@ -0,0 +1,299 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:test/test.dart' hide test, throws;
+import '../lib/realm.dart';
+import 'test.dart';
+
+Future<void> main([List<String>? args]) async {
+  print("Current PID $pid");
+
+  await setupTests(args);
+
+  test('Realm.syncSession throws on wrong configuration', () {
+    final config = Configuration.local([Task.schema]);
+    final realm = getRealm(config);
+    expect(() => realm.syncSession, throws<RealmError>());
+  });
+
+  baasTest('Realm.syncSession returns on FLX configuration', (configuration) async {
+    final realm = await getIntegrationRealm();
+
+    expect(realm.syncSession, isNotNull);
+    expect(realm.syncSession.path, realm.config.path);
+    expect(realm.syncSession, realm.syncSession);
+  });
+
+  baasTest('Realm.syncSession returns on FLX configuration', (configuration) async {
+    final realm = await getIntegrationRealm();
+
+    expect(realm.syncSession, isNotNull);
+    expect(realm.syncSession.path, realm.config.path);
+    expect(realm.syncSession, realm.syncSession);
+  });
+
+  baasTest('SyncSession.user returns a valid user', (configuration) async {
+    final app = App(configuration);
+    final user = await getIntegrationUser(app);
+    final config = Configuration.sync(user, [Task.schema]);
+    final realm = getRealm(config);
+
+    expect(realm.syncSession.user, user);
+    expect(realm.syncSession.user.id, user.id);
+  });
+
+  baasTest('SyncSession when isolate is torn down does not crash', (configuration) async {
+    final app = App(configuration);
+    final user = await getIntegrationUser(app);
+    final config = Configuration.sync(user, [Task.schema]);
+
+    // Don't use getRealm because we want the Realm to survive
+    final realm = Realm(config);
+
+    expect(realm.syncSession, isNotNull);
+  }, skip: 'crashes');
```

Comment:
I'll need help to figure out why this is crashing.

Reply:
```
===== CRASH =====
si_signo=Segmentation fault: 11(11), si_code=1, si_addr=0x1f0
version=2.17.0 (stable) (Mon May 9 10:36:47 2022 +0200) on "macos_x64"
pid=45917, thread=20227, isolate_group=main(0x7fdc21809e00), isolate=(nil)(0x0)
isolate_instructions=100a00020, vm_instructions=100a00020
  pc 0x00007ff816c09882 fp 0x00007000034d6b60 pthread_mutex_lock+0x4
  pc 0x00007ff816b6e6b9 fp 0x00007000034d6b70 std::__1::mutex::lock()+0x9
  pc 0x000000010c0b02eb fp 0x00007000034d6bc0 realm::SyncManager::unregister_session(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)+0x27
  pc 0x000000010c0b9104 fp 0x00007000034d6c00 realm::SyncSession::did_drop_external_reference()+0x7a
  pc 0x000000010c0bd606 fp 0x00007000034d6c20 realm::SyncSession::ExternalReference::~ExternalReference()+0x12
  pc 0x000000010c0b2581 fp 0x00007000034d6c40 std::__1::shared_ptr<realm::SyncSession>::~shared_ptr()+0x31
  pc 0x000000010be61116 fp 0x00007000034d6c60 realm_sync_session::~realm_sync_session()+0x12
  pc 0x0000000101131f87 fp 0x00007000034d6ca0 dart::FinalizablePersistentHandle::Finalize(dart::IsolateGroup*, dart::FinalizablePersistentHandle*)+0x67
  pc 0x0000000100b95e9d fp 0x00007000034d6d10 dart::Isolate::LowLevelCleanup(dart::Isolate*)+0x26d
  pc 0x0000000100b9720e fp 0x00007000034d6de0 dart::Isolate::Shutdown()+0x19e
  pc 0x0000000101136640 fp 0x00007000034d6e60 Dart_ShutdownIsolate+0xb0
  pc 0x0000000100bbf9ab fp 0x00007000034d6eb0 dart::MessageHandler::TaskCallback()+0x31b
  pc 0x0000000100cf4e87 fp 0x00007000034d6f30 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x147
  pc 0x0000000100cf52ee fp 0x00007000034d6f60 dart::ThreadPool::Worker::Main(unsigned long)+0x6e
  pc 0x0000000100c6331f fp 0x00007000034d6fb0 dart::OSThread::GetMaxStackSize()+0xaf
  pc 0x00007ff816c0e4e1 fp 0x00007000034d6fd0 _pthread_start+0x7d
  pc 0x00007ff816c09f6b fp 0x00007000034d6ff0 thread_start+0xf
-- End of DumpStackTrace
fish: Job 1, 'dart test --name crash test/s...' terminated by signal SIGABRT (Abort)
```
This happens because the finalization calls back into dart on a non-isolate thread (the isolate group still exists). I'll need to think a bit about this.

Reply:
I believe the issue is that during shutdown the order of finalizers is unpredictable. In particular in this test: `realm_app::~realm_app()` which destroys the `SyncManager` is called before `realm_sync_session::~realm_sync_session()` which tries to unregister from said manager.

This is illustrated by the following lldb session:
```
(lldb) bt
* thread #5, name = 'DartWorker', stop reason = step out
  * frame #0: 0x000000011c4343b9 librealm_dart.dylib`realm::SyncSession::become_inactive(realm::util::CheckedUniqueLock, std::__1::error_code) + 199
    frame #1: 0x000000011c43514f librealm_dart.dylib`realm::SyncSession::shutdown_and_wait() + 73
    frame #2: 0x000000011c4350e3 librealm_dart.dylib`realm::SyncSession::detach_from_sync_manager() + 15
    frame #3: 0x000000011c42eb0e librealm_dart.dylib`realm::SyncManager::~SyncManager() + 396
    frame #4: 0x000000011c40e2f3 librealm_dart.dylib`std::__1::shared_ptr<realm::SyncManager>::~shared_ptr() + 49
    frame #5: 0x000000011c405ecf librealm_dart.dylib`realm::app::App::~App() + 43
    frame #6: 0x000000011c40d999 librealm_dart.dylib`std::__1::shared_ptr<realm::app::App>::~shared_ptr() + 49
    frame #7: 0x000000011c1c77b8 librealm_dart.dylib`realm_app::~realm_app() + 18
    frame #8: 0x000000011028b7f7 dart`dart::FinalizablePersistentHandle::Finalize(dart::IsolateGroup*, dart::FinalizablePersistentHandle*) + 103
    frame #9: 0x000000010fce676d dart`dart::Isolate::LowLevelCleanup(dart::Isolate*) + 621
    frame #10: 0x000000010fce7e90 dart`dart::Isolate::Shutdown() + 448
    frame #11: 0x000000011028ff1b dart`Dart_ShutdownIsolate + 203
    frame #12: 0x000000010fd1087b dart`dart::MessageHandler::TaskCallback() + 795
    frame #13: 0x000000010fe49397 dart`dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*) + 327
    frame #14: 0x000000010fe497fe dart`dart::ThreadPool::Worker::Main(unsigned long) + 110
    frame #15: 0x000000010fdb471f dart`___lldb_unnamed_symbol5884$$dart + 159
    frame #16: 0x00007ff816c0e4e1 libsystem_pthread.dylib`_pthread_start + 125
    frame #17: 0x00007ff816c09f6b libsystem_pthread.dylib`thread_start + 15
(lldb) cont
Process 65421 resuming
Process 65421 stopped
* thread #5, name = 'DartWorker', stop reason = breakpoint 2.1
    frame #0: 0x000000011c430324 librealm_dart.dylib`realm::SyncManager::unregister_session(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
librealm_dart.dylib`realm::SyncManager::unregister_session:
->  0x11c430324 <+0>: pushq  %rbp
    0x11c430325 <+1>: movq   %rsp, %rbp
    0x11c430328 <+4>: pushq  %r15
    0x11c43032a <+6>: pushq  %r14
Target 0: (dart) stopped.
(lldb) bt
* thread #5, name = 'DartWorker', stop reason = breakpoint 2.1
  * frame #0: 0x000000011c430324 librealm_dart.dylib`realm::SyncManager::unregister_session(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
    frame #1: 0x000000011c438fb4 librealm_dart.dylib`realm::SyncSession::did_drop_external_reference() + 122
    frame #2: 0x000000011c43d466 librealm_dart.dylib`realm::SyncSession::ExternalReference::~ExternalReference() + 18
    frame #3: 0x000000011c4323e1 librealm_dart.dylib`std::__1::shared_ptr<realm::SyncSession>::~shared_ptr() + 49
    frame #4: 0x000000011c1e0cc3 librealm_dart.dylib`realm_sync_session::~realm_sync_session() + 143
    frame #5: 0x000000011c1e0b20 librealm_dart.dylib`realm_sync_session::~realm_sync_session() + 14
    frame #6: 0x000000011028b7f7 dart`dart::FinalizablePersistentHandle::Finalize(dart::IsolateGroup*, dart::FinalizablePersistentHandle*) + 103
    frame #7: 0x000000010fce676d dart`dart::Isolate::LowLevelCleanup(dart::Isolate*) + 621
    frame #8: 0x000000010fce7e90 dart`dart::Isolate::Shutdown() + 448
    frame #9: 0x000000011028ff1b dart`Dart_ShutdownIsolate + 203
    frame #10: 0x000000010fd1087b dart`dart::MessageHandler::TaskCallback() + 795
    frame #11: 0x000000010fe49397 dart`dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*) + 327
    frame #12: 0x000000010fe497fe dart`dart::ThreadPool::Worker::Main(unsigned long) + 110
    frame #13: 0x000000010fdb471f dart`___lldb_unnamed_symbol5884$$dart + 159
    frame #14: 0x00007ff816c0e4e1 libsystem_pthread.dylib`_pthread_start + 125
    frame #15: 0x00007ff816c09f6b libsystem_pthread.dylib`thread_start + 15
(lldb)
```

Reply:
I think this is the second situation where we really need to control the order of finalization.

Reply:
Perhaps the time has come to revisit https://github.com/realm/realm-dart/pull/528. Now we also have weak references with dart 2.17.

Reply:
@danieltabacaru Could you take at look at this?

Reply:
After discussing with @danieltabacaru I have opened an issue with core.